apis/r/R/ReadIter.R

```diff
@@ -0,0 +1,84 @@
+#' SOMA Read Iterator Base class
+#'
+#' Class that allows for read iteration of SOMA reads.
+
+ReadIter <- R6::R6Class(
+  classname = "ReadIter",
+
+  public = list(
+                
+    #' @description Create (lifecycle: experimental)
+    #' @param uri Character value with URI path to a SOMADataFrame or SOMASparseNDArray
+    #' @param config character vector containing TileDB config.
+    #' @param colnames Optional vector of character value with the name of the columns to retrieve
+    #' @param qc Optional external Pointer object to TileDB Query Condition, defaults to \sQuote{NULL} i.e.
+    #' no query condition
+    #' @param dim_points Optional named list with vector of data points to select on the given
+    #' dimension(s). Each dimension can be one entry in the list.
+    #' @param loglevel Character value with the desired logging level, defaults to \sQuote{auto}
+    #' which lets prior setting prevail, any other value is set as new logging level.
+    initialize = function(uri, 
+                          config, 
+                          colnames = NULL, 
+                          qc = NULL, 
+                          dim_points = NULL, 
+                          loglevel = "auto") {
+        
+        # Instantiate soma_reader_pointer with a soma_array_reader object
+          private$soma_reader_pointer <- sr_setup(
+            uri = uri,
+            config = config,
+            colnames = colnames,
+            qc = qc,
+            dim_points = dim_points,
+            loglevel = loglevel
+          )
+    },
+
+    #' @description Check if iterated read is complete or not. (lifecycle: experimental)
+    #' @return logical
+    read_complete = function() {
+      if (is.null(private$soma_reader_pointer)) {
+          TRUE
+      } else {
+          sr_complete(private$soma_reader_pointer)
+      }
+    },
+
+    #' @description Read the next chunk of an iterated read. (lifecycle: experimental).
+    #' If read is complete, retunrs `NULL` and raises warning.
+    #' @return \code{NULL} or one of arrow::\link[arrow]{Table}, \link{matrixZeroBasedView}
+    read_next = function() {
+      if (is.null(private$soma_reader_pointer)) {
+          NULL
+      } else {
+          if (self$read_complete()) {
+              warning("Iteration complete, returning NULL")
+              NULL
+          } else {
+              rl <- sr_next(private$soma_reader_pointer)
+              return(private$soma_reader_transform(rl))
+          }
+      }
+    },
+    
+    #' @description  Concatenate remainder of iterator
+    # to be refined in derived classes
+    concat = function() {
```

Comment:
Sorry for the very belated review -- @aaronwolen asked me to help review, and I'm happy to -- I believe my review is speaking for the both of us

There are definitely multiple ways to go here. However, given that on the Python side we have a single `read`, options of `.tables`, etc., and `concat` -- e.g. `sdf.read().tables().concat()` -- we should do the same here. It won't be ideal for everyone but it will work, it will parallel the Python implementation, and we can add some keystroke-saving syntactic-sugar functions _on top of these_ later if we want.

Let's get this `concat` going, and get rid of the `iterated` argument to `read`. Then `read` will always return an iterator (low complexity), and `tables()` et al. will transform an iterator of one type to an iterator of another, and `concat` will have the single job of doing concatenation.

Note that `apis/r/R/utils-readerTransformers.R` on this PR already has reader-transformers, so (I think and hope) it should be easy to connect them as methods on the iterator objects.

Likewise, `concat()` exists in non-stub form elsewhere on this PR and should be able to be connected as a method on the iterator objects.

If `read_next` remains, it needs to be remain as a helper method, not as public API.

Reply:
Agreed with everything and I will try to get this done tonight.

> If read_next remains, it needs to be remain as a helper method, not as public API.

@johnkerl I'm not sure that the suggestion here is? Unfortunately in R there's no generic `next()` function, so I'm not sure how we could give users the ability to iterate themselves unless we:

1. make `read_next` a public method
2. OR we create function `next()` that calls internally `read_next`

The second one is a little less intuitive imo.

Reply:
> read will always return an iterator

I'm not sure I understand the benefit to this. While R does have [package that implements iterators](https://cran.r-project.org/package=iterators), it isn't widely used (or at all in the single-cell world). Because iterators don't exist in R, all this does is add complexity to R users. While I understand the desire to make the R API to act like the Python API, we do need to keep in mind that R is not Python and there are things that Python handles natively that just don't work as well in R

Reply:
I think the key is here:

> we can add some keystroke-saving syntactic-sugar functions on top of these later if we want

and imo it's not a "if we want" but more a "we need to". I think we can keep the low-level functionality in the R6 level as proposed and then we provide wrapper functions like `as_arrow_table(soma_obj)`[ see here](https://arrow.apache.org/docs/dev/r/reference/as_arrow_table.html), `as(soma_obj, "dgCMatrix")`,  `as.data.frame(soma_obj)`, etc

Reply:
@johnkerl I have addressed your comment with the exception of 

> If read_next remains, it needs to be remain as a helper method, not as public API.

Reply:
@pablo-gar looking this morning -- thank you! :)

Reply:
@pablo-gar sorry to confuse.

This looks good to me -- thank you!! :)

Reply:
Awesome! I will update the docs and merge main into the PR branch